File: leetcode/easy/1608. Special Array With X Elements Greater Than or Equal X/main.go
package main

import "sort"

func specialArray(nums []int) int {
	sort.Ints(nums)

	start, end := 0, len(nums)

	for start <= end {
		mid := start + (end-start)/2
		freq := count(nums, mid)

		if freq == mid {
			return mid
		} else if freq < mid {
			end = mid - 1
		} else {
			start = mid + 1
		}

	}

	return -1
}

func count(nums []int, target int) int {
	counter := 0

	for _, v := range nums {
		if v >= target {
			counter++

		}
	}

	return counter
}

File: leetcode/easy/1608. Special Array With X Elements Greater Than or Equal X/main.rs
impl Solution {
    pub fn special_array(nums: Vec<i32>) -> i32 {
        let mut nums = nums.clone();
        nums.sort();

        let (mut start, mut end) = (0, nums.len());

        while start <= end {
            let mid = start + (end - start) / 2;
            let freq = Self::count(&nums, mid as i32);

            if mid == freq {
                return mid as i32;
            } else if freq > mid {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        -1
    }

    fn count(nums: &Vec<i32>, target: i32) -> usize {
        let mut counter = 0;

        for &el in nums {
            if el >= target {
                counter += 1;
            }
        }

        counter
    }
}
